src/tbbmalloc/backend.cpp

```diff
@@ -317,11 +319,19 @@ inline bool BackendSync::waitTillBlockReleased(intptr_t startModifiedCnt)
         if (currCoalescQInFlyBlocks > 0 && backend->scanCoalescQ(/*forceCoalescQDrop=*/false))
             break;
         // 4) when there are no blocks
-        if (!currBinsInFlyBlocks && !currCoalescQInFlyBlocks)
+        if (!currBinsInFlyBlocks && !currCoalescQInFlyBlocks) {
             // re-scan make sense only if bins were modified since scanned
+            auto pool = backend->extMemPool;
+            if (pool->hardCachesCleanupInProgress || pool->softCachesCleanupInProgress) {
+                backoff.pause();
+                continue;
+            }
+
             return startModifiedCnt != getNumOfMods();
```

Comment:
What if `cleanup` starts after you checked the condition (line 325) but before executing this `return` statement? Is this OK?

Reply:
I don't care - in theory we could wait infinitely, until someone will free memory for us.
What is important is to not return ENOMEM while there was a free memory when user called malloc()

So in your case, if we are in this line - minimum one cleanup* has finished (it was done by this thread, or other)


* for smaller allocation waitTillBlockReleased is called twice (ones before cleanup, and second time after cleanup) so after first call, this sentence is not true - but it doesn't matter anyway